include/flatbuffers/flatbuffers.h

```diff
@@ -1768,22 +1795,43 @@ class FlatBufferBuilder {
   /// where the vector is stored.
   template<typename T, typename F, typename S>
   Offset<Vector<T>> CreateVector(size_t vector_size, F f, S *state) {
+    FLATBUFFERS_ASSERT(FLATBUFFERS_GENERAL_HEAP_ALLOC_OK);
     std::vector<T> elems(vector_size);
     for (size_t i = 0; i < vector_size; i++) elems[i] = f(i, state);
     return CreateVector(elems);
   }
 
   /// @brief Serialize a `std::vector<std::string>` into a FlatBuffer `vector`.
-  /// This is a convenience function for a common case.
+  /// This is a convenience function for a common case. This may use a vector
+  /// stored on the heap to cache the intermediate results, but only stores the
+  /// numerical offsets.
   /// @param v A const reference to the `std::vector` to serialize into the
   /// buffer as a `vector`.
   /// @return Returns a typed `Offset` into the serialized data indicating
   /// where the vector is stored.
   Offset<Vector<Offset<String>>> CreateVectorOfStrings(
       const std::vector<std::string> &v) {
-    std::vector<Offset<String>> offsets(v.size());
-    for (size_t i = 0; i < v.size(); i++) offsets[i] = CreateString(v[i]);
-    return CreateVector(offsets);
+    size_t scratch_buffer_usage = v.size() * sizeof(Offset<String>);
+    if (scratch_buffer_usage < buf_.scratch_size()) {
+      // There is enough room in the scratch buffer to store all the string
+      // offsets, use it instead of creating the temporary buffer on the heap.
+      for (size_t i = 0; i < v.size(); i++) {
+        buf_.scratch_push_small(CreateString(v[i]));
+      }
+      StartVector(v.size(), sizeof(Offset<String>));
+      for (auto it = buf_.scratch_end();
+           it > buf_.scratch_end() - scratch_buffer_usage;) {
+        it -= sizeof(Offset<String>);
+        PushElement(*reinterpret_cast<Offset<String> *>(it));
+      }
+      buf_.scratch_pop(scratch_buffer_usage);
+      return Offset<Vector<Offset<String>>>(EndVector(v.size()));
+    } else {
+      FLATBUFFERS_ASSERT(FLATBUFFERS_GENERAL_HEAP_ALLOC_OK);
+      std::vector<Offset<String>> offsets(v.size());
+      for (size_t i = 0; i < v.size(); i++) offsets[i] = CreateString(v[i]);
+      return CreateVector(offsets);
+    }
```

Comment:
This method might be generalized, for example:
```cpp
template<class StringContainer>
Offset<Vector<Offset<String>>> CreateVectorOfStrings(const StringContainer &v) {
     for (auto it=v.cbegin(); it != v.cend(); ++it) {
        buf_.scratch_push_small(CreateString(*it));
      }
}
```
or even better:
```cpp
template<class It>
Offset<Vector<Offset<String>>> CreateVectorOfStrings(It begin, It end) {
     for (auto it=begin; it != end; ++it) {
        buf_.scratch_push_small(CreateString(*it));
      }
}
```
The second version can accept any container of:  `const char*`, `std::string`, `std::string_vew`.

Reply:
Done.

Reply:
Very cool idea to use the scratch buffer here, though that is only essential for the non-general-alloc users. For everyone else, this could be a speedup, but creating a vector of strings is a pretty heavy operation anyway, so I am not sure if the removal of `std::vector` is any speedup, and worth adding the extra code complexity. Opinions @vglavnyy ?

Reply:
It's more of avoiding the heap than speed up concerns. It would be nice to just use the scratch buffer wherever we would use the heap though, if we could add some internal APIs to ensure space.

It would also help remove some std:: libraries from the code to enable more embedded platforms as discussed in #6157

Another idea is to just make some overloaded methods that take in a container class to use as the temporary storage. That way the user can decide what data structures are being used under the hood.

Reply:
ok, that is generally a good direction, lets keep it.

Reply:
Because this is storing `Offset<String>`, I feel like it is OK to store them all in the scratch buffer region, as any increase in buffer allocation this would cause is dwarfed by the creation of the strings themselves. So it feels safe to completely forgo the vector implementation in this case.